examples/step-63/step-63.cc

```diff
@@ -19,16 +19,1091 @@
+              ordered_indices = create_downstream_order(dof_handler,
+                                                        advection_direction,
+                                                        level);
+            else if (settings.dof_renum == "upstream")
+              ordered_indices =
+                create_downstream_order(dof_handler,
+                                        -1.0 * advection_direction,
+                                        level);
+            else if (settings.dof_renum == "random")
+              ordered_indices = create_random_order(dof_handler, level);
+            else if (settings.dof_renum == "none")
+              {
+                // Do nothing
+              }
+            else
+              AssertThrow(false, ExcNotImplemented());
+
+            smoother_data[level].order =
+              std::vector<std::vector<unsigned int>>(1, ordered_indices);
+          }
+
+        auto smoother =
+          std_cxx14::make_unique<MGSmootherPrecondition<SparseMatrix<double>,
+                                                        Smoother,
+                                                        Vector<double>>>();
+        smoother->initialize(mg_matrices, smoother_data);
+        smoother->set_steps(2);
+        return smoother;
+      }
+    else
+      AssertThrow(false, ExcNotImplemented());
+  }
+
+
+  template <int dim>
+  void AdvectionProblem<dim>::solve()
+  {
+    Timer time;
+
+    const double       solve_tol = 1e-8 * system_rhs.l2_norm();
+    const unsigned int max_iters = 200;
+    SolverControl      solver_control(max_iters, solve_tol, true, true);
+    solver_control.enable_history_data();
+
+    typedef MGTransferPrebuilt<Vector<double>> Transfer;
+    Transfer                                   mg_transfer(mg_constrained_dofs);
+    mg_transfer.build_matrices(dof_handler);
+
+    FullMatrix<double> coarse_matrix;
+    coarse_matrix.copy_from(mg_matrices[0]);
+    MGCoarseGridHouseholder<double, Vector<double>> coarse_grid_solver;
+    coarse_grid_solver.initialize(coarse_matrix);
+
+    std::unique_ptr<MGSmoother<Vector<double>>> mg_smoother = create_smoother();
+
+    mg::Matrix<Vector<double>> mg_matrix(mg_matrices);
+    mg::Matrix<Vector<double>> mg_interface_matrix_in(mg_interface_in);
+    mg::Matrix<Vector<double>> mg_interface_matrix_out(mg_interface_out);
+
+    Multigrid<Vector<double>> mg(
+      mg_matrix, coarse_grid_solver, mg_transfer, *mg_smoother, *mg_smoother);
+    mg.set_edge_matrices(mg_interface_matrix_out, mg_interface_matrix_in);
+
+    PreconditionMG<dim, Vector<double>, Transfer> preconditioner(dof_handler,
+                                                                 mg,
+                                                                 mg_transfer);
+
+
+    std::cout << "     Solving with GMRES to tol " << solve_tol << "..."
+              << std::endl;
+    SolverGMRES<> solver(solver_control);
+
+    time.restart();
```

Comment:
There is no loop in this function, so just move the declaration of the `time` variable right here and call `time.start()` instead.